proposals/native-lock.md

```diff
@@ -0,0 +1,121 @@
+# Native lock
+
+## Summary
+[summary]: #summary
+
+Special-case how `System.Threading.Lock` interacts with the `lock` keyword (calling its `EnterLockScope` method under the hood).
+Add static analysis warnings to prevent accidental misuse of the type where possible.
+
+## Motivation
+[motivation]: #motivation
+
+.NET 9 is introducing a new [`System.Threading.Lock` type](https://github.com/dotnet/runtime/issues/34812)
+as a better alternative to existing monitor-based locking.
+The presence of the `lock` keyword in C# might lead developers to think they can use it with this new type.
+Doing so wouldn't lock according to the semantics of this type but would instead treat it as any other object and would use monitor-based locking.
+
+## Detailed design
+[design]: #detailed-design
+
+Semantics of the lock statement ([§13.13](https://github.com/dotnet/csharpstandard/blob/9af5bdaa7af535f34fbb7923e5406e01db8489f7/standard/statements.md#1313-the-lock-statement))
+are changed to special-case the `System.Threading.Lock` type:
+
+> A `lock` statement of the form `lock (x) { ... }`
+>
+> 1. **where `x` is an expression of type `System.Threading.Lock`, is precisely equivalent to:**
+>    ```cs
+>    using (x.EnterLockScope())
+>    {
+>        ...
+>    }
+>    ```
+>    **and `System.Threading.Lock` must have the following shape:**
+>    ```cs
+>    namespace System.Threading
+>    {
+>        class Lock
+>        {
+>            public Scope EnterLockScope();
```

Comment:
FYI I updated the proposal in the [Lock API review](https://github.com/dotnet/runtime/issues/34812#issuecomment-1898861386) to rename this method back to `EnterScope`, as the `Lock` in the name is redundant given the type name and it's not being suggested for use as a general pattern anymore.

Reply:
> it's not being suggested for use as a general pattern anymore

I think that's still considered as future work (when ref structs can implement interfaces). Wouldn't it be better to keep the name `EnterLockScope` to make the transition to the generic pattern smoother?

Reply:
I was under the impression that it may or may not be expanded to a general lock pattern, and figured we might as well use an appropriate name for now. If a general pattern is exposed it may also be more useful for it to be usable by other existing types, which may also involve new syntax or some such. At that time if it's necessary to add `EnterLockScope` again we could, and perhaps explicitly implement it if it's an interface member to hide it.

Reply:
OK, I assume this can still change in API review. Do you know when the API will be approved?

Reply:
Yep, I don't think it's been scheduled yet, will add you when I find out

Reply:
The spec should reflect our best guess for what the API will be. If API review changes the name we will just update the spec. 

Based on previous experiences though we should put the name @kouvel least wants in the spec. Cause inevitably whatever we pick in the initial spec always changes :wink: